### PR TITLE
Add Process Reward Model.

### PR DESCRIPTION
**Pull request by XuLingnan**
### PR types

New features
### PR changes

Models
### Description

Support step-wise PRM training for Llama & Mistral models.
- DataFlow: 
{
    "src" : ["prompt"],
    "tgt" : [],
    "responses" : ["step_1", ..., "step_k"]
    "labels" : ["label_1", ..., "label_k"]
}
- Model: Calculate CE loss at the end of each PRM step.
- Training process: Support PRM training & arguments (--process_reward, --placeholder_token, --reward_tokens).

### CODE REVIEW

File: llm/alignment/rm/flashmask/data.py [hunk 1]
```diff
@@ -163,3 +164,187 @@ def preference_collate_fn(batch, max_seq_len=None):
...
+    
+    # TODO:
+    response_token_ids = tokenizer(
+        model_args.placeholder_token.join(
+            # [" "+local_data+" " if idx==0 else local_data+" " for idx, local_data in enumerate(data["responses"])]
+            data["responses"]
+        ) + model_args.placeholder_token,
+        add_special_tokens=True,
+    )["input_ids"]
+    # print(repr(
```
Comment (gongel) [inline llm/alignment/rm/flashmask/data.py]: 所有的debug打印信息可以去掉

File: llm/alignment/rm/flashmask/reward_model.py [hunk 2]
```diff
@@ -153,4 +154,130 @@ def _get_name_mappings(cls, config: LlamaConfig) -> list[StateDictNameMapping]:
         return mappings
 
 
+class LlamaModelForProcessScore(LlamaPretrainedModel):
```
Comment (gongel) [inline llm/alignment/rm/flashmask/reward_model.py]: 只能支持llama模型吗？其他的模型能否支持

File: llm/alignment/rm/flashmask/reward_model.py [hunk 3]
```diff
@@ -153,4 +154,130 @@ def _get_name_mappings(cls, config: LlamaConfig) -> list[StateDictNameMapping]:
...
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            attn_mask_startend_row_indices=attn_mask_startend_row_indices,
+        )
+        
+        placeholder_mask = input_ids == self.placeholder_token_id
+        if return_dict:
+            logits = outputs.logits
+        elif labels is not None:
+            # import pdb; pdb.set_trace()
```
Comment (gongel) [inline llm/alignment/rm/flashmask/reward_model.py]: pdb都可以删掉

File: llm/alignment/rm/flashmask/reward_model.py [hunk 4]
```diff
@@ -153,4 +154,130 @@ def _get_name_mappings(cls, config: LlamaConfig) -> list[StateDictNameMapping]:
...
+            logits = outputs[0]
+        # breakpoint()
+        logits = logits[placeholder_mask]
+        labels = labels[placeholder_mask]
+        # breakpoint()
+        logits = logits[..., self.reward_token_ids]
+        for idx, token in enumerate(self.reward_token_ids):
+            labels = paddle.where(labels == token, idx, labels)
+
+        loss = self.loss(logits, labels)
```
Comment (gongel) [inline llm/alignment/rm/flashmask/reward_model.py:310]: 这里的loss计算是否需要考虑黑白名单

**Review by gongel — COMMENTED** [review]
缺少readme和单测，可参考https://github.com/PaddlePaddle/PaddleNLP/blob/develop/llm/docs/rlhf.md

**Comment by gongel** [comment]
CLA也需要签署，否则无法合PR

File: llm/alignment/rm/flashmask/run_reward.py [hunk 5]
```diff
@@ -19,9 +19,12 @@
 from functools import partial
 
 import paddle
-from data import preference_collate_fn, preprocess_preference_data
+from data import (
+    preference_collate_fn, preprocess_preference_data, 
+    process_collate_fn, preprocess_process_data
+)
 from reward_argument import DataArgument, ModelArgument, TrainingArguments
```
Comment (KB-Ding) [inline llm/alignment/rm/flashmask/run_reward.py:22]: 写法中没有考虑TrainingArguments 的sequence_parallel，应该暂时强制关掉，目前看如果传入为True，会有报错。

File: llm/config/llama/prm_flashmask_argument.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
+{
+    "model_name_or_path": "meta-llama/Meta-Llama-3-8B",
+    "train_dataset_path": "./data/Math-Shepherd/test.json",
+    "dev_dataset_path": "./data/Math-Shepherd/test.json",
```
Comment (KB-Ding) [inline llm/config/llama/prm_flashmask_argument.json:4]: 数据集来源是否应该写个README？并附上样例数据链接。
另外数据格式也需要说明？

File: llm/alignment/rm/flashmask/reward_trainer.py [hunk 7]
```diff
@@ -45,15 +53,28 @@ def get_batch_metrics(self, model, batch, train_eval="train"):
...
-        for key in metrics:
-            metrics[key] = self._nested_gather(paddle.tile(metrics[key], repeat_times=[1, 1])).mean().cpu()
-        if self.args.should_save:
-            self.store_metrics(metrics, train_eval=train_eval)
-        return loss
+        if not self.process_reward:
+            loss, chosen_scores, rejected_scores = model(**rm_inputs)
+            prefix = "eval_" if train_eval == "eval" else ""
+            metrics = {}
+            metrics[f"{prefix}accuracy"] = (chosen_scores > rejected_scores).astype("float32").mean()
```
Comment (KB-Ding) [inline llm/alignment/rm/flashmask/reward_trainer.py:60]: 看起来在这里额外分支一下就行，不需要在另一个分支复制这么多代码？

File: llm/alignment/rm/flashmask/reward_trainer.py [hunk 8]
```diff
@@ -45,15 +53,28 @@ def get_batch_metrics(self, model, batch, train_eval="train"):
...
+            metrics = {}
+            metrics[f"{prefix}accuracy"] = accs.astype("float32").mean()
+            for key in metrics:
+                metrics[key] = self._nested_gather(paddle.tile(metrics[key], repeat_times=[1, 1])).mean().cpu()
+            if self.args.should_save:
+                self.store_metrics(metrics, train_eval=train_eval)
+            return loss
+            
 
     def compute_loss(self, model, inputs):
```
Comment (KB-Ding) [inline llm/alignment/rm/flashmask/reward_trainer.py]: 这里与非flashmask的RM计算区别有些多，这里是否需要补齐？
https://github.com/PaddlePaddle/PaddleNLP/blob/develop/llm/alignment/rm/reward_trainer.py#L106-L186，

File: llm/alignment/rm/flashmask/reward_trainer.py [hunk 9]
```diff
@@ -24,19 +24,27 @@ class RewardTrainer(Trainer):
     Initialize RewardTrainer.
     """
 
-    def __init__(self, model, data_collator, **kwargs):
+    def __init__(self, model, data_collator, process_reward=False, **kwargs):
         super().__init__(model, data_collator=data_collator, **kwargs)
         self._stored_metrics = defaultdict(lambda: defaultdict(list))
         if self.compute_metrics is not None:
             raise NotImplementedError("compute_metrics is not supported for RewardTrainer")
```
Comment (KB-Ding) [inline llm/alignment/rm/flashmask/reward_trainer.py:31]: 这里有个风险，trainer的所有call_back都拿不到compute_metrics的返回，可能导致eval指标打不出来？
看看是否能测试下eval是否有问题，以及看看能否补齐？https://github.com/PaddlePaddle/PaddleNLP/blob/develop/llm/alignment/rm/reward_trainer.py#L89-L90

File: llm/alignment/rm/flashmask/data.py [hunk 10]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
```
Comment (KB-Ding) [inline llm/alignment/rm/flashmask/data.py:13]: 代码整体清理一下无用的注释

File: llm/config/llama/prm_flashmask_argument.json [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    "output_dir": "./checkpoints/prm_ckpts",
+    "process_reward": true,
+    "placeholder_token": "ки",
```
Comment (KB-Ding) [inline llm/config/llama/prm_flashmask_argument.json:7]: 这个是否与模型绑定？应该说明下为什么是这个token

File: llm/config/llama/prm_flashmask_argument.json [hunk 12]
```diff
@@ -0,0 +1,39 @@
+{
+    "model_name_or_path": "meta-llama/Meta-Llama-3-8B",
```
Comment (KB-Ding) [inline llm/config/llama/prm_flashmask_argument.json:2]: COT数据长度如何？8B情况下显存占用大概在什么级别？目前的写法分布式支持比较少，需要考虑下一步的分布式优化方向。

**Comment by KB-Ding** [comment]
> ### PR types
> ### PR changes
> ### Description
补充一下PR的描述

**Comment by gongel** [comment]
精度是否和竞品做过对齐？

**Comment by gongel** [comment]
LLM目录下增加文档，凸显PRM

**Review by DrownFish19 — APPROVED** [review]
LGTM
